src/source/PeerConnection/Rtp.c

```diff
@@ -181,19 +181,27 @@ STATUS writeFrame(PRtcRtpTransceiver pRtcRtpTransceiver, PFrame pFrame)
     CHK_STATUS(constructRtpPackets(pPayloadArray, pKvsRtpTransceiver->sender.payloadType, pKvsRtpTransceiver->sender.sequenceNumber, rtpTimestamp, pKvsRtpTransceiver->sender.ssrc, pPacketList, pPayloadArray->payloadSubLenSize));
     pKvsRtpTransceiver->sender.sequenceNumber = GET_UINT16_SEQ_NUM(pKvsRtpTransceiver->sender.sequenceNumber + pPayloadArray->payloadSubLenSize);
 
+    bufferAfterEncrypt = (pKvsRtpTransceiver->sender.payloadType == pKvsRtpTransceiver->sender.rtxPayloadType);
     for (i = 0; i < pPayloadArray->payloadSubLenSize; i++) {
         pRtpPacket = pPacketList + i;
         CHK_STATUS(createBytesFromRtpPacket(pRtpPacket, &rawPacket, &packetLen));
-        rawLen = packetLen;
 
+        if (!bufferAfterEncrypt) {
+            pRtpPacket->pRawPacket = rawPacket;
+            pRtpPacket->rawPacketLength = packetLen;
+            CHK_STATUS(rtpRollingBufferAddRtpPacket(pKvsRtpTransceiver->sender.packetBuffer, pRtpPacket));
+        }
+
+        signedPacketLen = packetLen;
         rawPacket = MEMREALLOC(rawPacket, packetLen + 10); // For SRTP authentication tag
-        pRtpPacket->pRawPacket = rawPacket;
-        pRtpPacket->rawPacketLength = packetLen;
-        if (pKvsRtpTransceiver->sender.packetBuffer != NULL) {
+        CHK_STATUS(encryptRtpPacket(pKvsPeerConnection->pSrtpSession, rawPacket, &signedPacketLen));
+        CHK_STATUS(iceAgentSendPacket(pKvsPeerConnection->pIceAgent, rawPacket, signedPacketLen));
+
+        if (bufferAfterEncrypt) {
+            pRtpPacket->pRawPacket = rawPacket;
+            pRtpPacket->rawPacketLength = signedPacketLen;
```

Comment:
Winder if the logic should be to simply check if the packet needs encryption in which case re-alloc and encrypt and set the new size. Set the size in pRtpPacket bookkeeping, add to rolling buffer and then ice agent send packet?